plugins/node/instrumentation-kafkajs/src/instrumentation.ts

```diff
@@ -221,47 +409,75 @@ export class KafkaJsInstrumentation extends InstrumentationBase<KafkaJsInstrumen
-        const receivingSpan = instrumentation._startConsumerSpan(
-          payload.batch.topic,
-          undefined,
-          MESSAGINGOPERATIONVALUES_RECEIVE,
-          ROOT_CONTEXT
-        );
+        const receivingSpan = instrumentation._startConsumerSpan({
+          topic: payload.batch.topic,
+          message: undefined,
+          operationType: MESSAGING_OPERATION_TYPE_VALUE_RECEIVE,
+          ctx: ROOT_CONTEXT,
+          attributes: {
+            [ATTR_MESSAGING_BATCH_MESSAGE_COUNT]: payload.batch.messages.length,
+            [ATTR_MESSAGING_DESTINATION_PARTITION_ID]: String(
+              payload.batch.partition
+            ),
+          },
+        });
         return context.with(
           trace.setSpan(context.active(), receivingSpan),
           () => {
-            const spans = payload.batch.messages.map(
-              (message: KafkaMessage) => {
-                const propagatedContext: Context = propagation.extract(
-                  ROOT_CONTEXT,
-                  message.headers,
-                  bufferTextMapGetter
-                );
-                const spanContext = trace
-                  .getSpan(propagatedContext)
-                  ?.spanContext();
-                let origSpanLink: Link | undefined;
-                if (spanContext) {
-                  origSpanLink = {
-                    context: spanContext,
-                  };
-                }
-                return instrumentation._startConsumerSpan(
-                  payload.batch.topic,
-                  message,
-                  MESSAGINGOPERATIONVALUES_PROCESS,
-                  undefined,
-                  origSpanLink
-                );
+            const startTime = Date.now();
+            const spans: Span[] = [];
+            const pendingMetrics: RecordPendingMetric[] = [];
+            payload.batch.messages.forEach(message => {
+              const propagatedContext: Context = propagation.extract(
+                ROOT_CONTEXT,
+                message.headers,
+                bufferTextMapGetter
+              );
+              const spanContext = trace
+                .getSpan(propagatedContext)
+                ?.spanContext();
+              let origSpanLink: Link | undefined;
+              if (spanContext) {
+                origSpanLink = {
+                  context: spanContext,
+                };
               }
-            );
+              spans.push(
+                instrumentation._startConsumerSpan({
+                  topic: payload.batch.topic,
+                  message,
+                  operationType: MESSAGING_OPERATION_TYPE_VALUE_PROCESS,
+                  link: origSpanLink,
+                  attributes: {
+                    [ATTR_MESSAGING_DESTINATION_PARTITION_ID]: String(
+                      payload.batch.partition
+                    ),
+                  },
+                })
+              );
```

Comment:
This isn't new code but it feels a little bit wrong when compared to the examples in the semantic conventions doc ([this one](https://opentelemetry.io/docs/specs/semconv/messaging/messaging-spans/#batch-receiving), specifically). All of these `process` spans for each message will have the same start/end times as the batch itself and I'm not sure that's helpful or intuitive. But for now, I'm leaving that design choice alone and just adding the corresponding metric that the semantic convention requires for all `process` spans.

Reply:
Yeah this might be out of scope for this PR. Think creating a single `receive` span with `messaging.batch.message_count` attribute would be fine and then add the message ID, parent span context as links, not as new spans. (Trying this out with SQS: https://github.com/open-telemetry/opentelemetry-js-contrib/pull/2345)